tests/unit/lms/validation/_exceptions_test.py

```diff
@@ -34,10 +34,9 @@ def test_we_produce_a_string_message(self, redirect):
         assert redirect.detail == Any.string.matching(".*field_name.*error_name.*")
 
     @pytest.mark.parametrize(
-        "description,messages",
-        [("string", "a string"), ("list", []), ("dict of string", {"a": "b"})],
+        "messages", ["a string", [], {"a": "b"}],
     )
-    def test_is_requires_well_formatted_messages(self, description, messages):
+    def test_it_requires_well_formatted_messages(self, messages):
```

Comment:
The above does stringify nicely when running the tests. Would this be worth changing to `_description`?

Reply:
pytest actually has a way of adding descriptions (which it calls "ids") to the tests, when you want these to actually be used as the test names in pytest so that they show up in pytest's output (as opposed to just putting code comments next to each parameter, which we often do, but obviously that doesn't do anything to pytest's output).

I'll change it to use pytest ids properly instead.

The way it works is that normally pytest auto-generates an id for each parametrized test case based on the param names and values, but it also provides a few different ways for you to set the ids yourself if you want to:

You can pass a list of id strings which must be in the same order as the test cases in `testdata`:

```python
@pytest.mark.parametrize("a,b,expected", testdata, ids=["forward", "backward"])
```

Alternatively you can use `pytest.param()` so that the ids are grouped together with the test data:

```python
@pytest.mark.parametrize(
    "a,b,expected",
    [
        pytest.param(
            datetime(2001, 12, 12), datetime(2001, 12, 11), timedelta(1), id="forward"
        ),
        pytest.param(
            datetime(2001, 12, 11), datetime(2001, 12, 12), timedelta(-1), id="backward"
        ),
    ],
)
```

Another alternative: you provide a function that generates the ID for each test case:

```python
def idfn(val):
    if isinstance(val, (datetime,)):
        # note this wouldn't show any hours/minutes/seconds
        return val.strftime("%Y%m%d")

@pytest.mark.parametrize("a,b,expected", testdata, ids=idfn)
def test_timedistance_v2(a, b, expected):
    ...
```

* * * 

Parametrize is actually a lot more powerful than I usually remember.
See <https://docs.pytest.org/en/latest/parametrize.html> and
<https://docs.pytest.org/en/latest/example/parametrize.html#deferring-the-setup-of-parametrized-resources>. Some examples:

Apply a mark to just one of the parametrized instances of a test. Here the
`(6*9, 42)` version of the test is marked as expected to fail:

```python
@pytest.mark.parametrize(
    "test_input,expected",
    [("3+5", 8), ("2+4", 6), pytest.param("6*9", 42, marks=pytest.mark.xfail)],
)
```

Similarly you can mark just one test case as expected to raise a specific
error:

```python
@pytest.mark.parametrize(
    "example_input,expectation",
    [
        (3, does_not_raise()),
        (2, does_not_raise()),
        (1, does_not_raise()),
        (0, pytest.raises(ZeroDivisionError)),
    ],
)
def test_division(example_input, expectation):
    ...
```

You can defer the parametrization of a test function, which normally happens at
test collection time, until test run time with `indirect=True` or
`indirect=["x"]` to defer the generation of the parametrized `x` argument only:
https://docs.pytest.org/en/latest/example/parametrize.html#deferring-the-setup-of-parametrized-resources

You can get pytest to generate the cross-product, e.g. this generates every
combination of the different values for x and y:

```python
@pytest.mark.parametrize("x", [0, 1])
@pytest.mark.parametrize("y", [2, 3])
def test_foo(x, y):
    pass
```

There's something called `pytest_generate_tests` that's for when you need to do
more powerful test generation: https://docs.pytest.org/en/latest/parametrize.html#basic-pytest-generate-tests-example

Reply:
I've changed this to:

```python
    @pytest.mark.parametrize(
        "messages",
        [
            pytest.param("a string", id="string"),
            pytest.param([], id="list"),
            pytest.param({"a": "b"}, id="dict of string"),
        ],
    )
    def test_it_requires_well_formatted_messages(self, messages):
        with pytest.raises(ValueError):
            LTIToolRedirect("http://example.com", messages)
```

Test output now:

```terminal
> .tox/tests/bin/pytest tests/unit/lms/validation/_exceptions_test.py -vv
...
tests/unit/lms/validation/_exceptions_test.py::TestLTIToolRedirect::test_it_requires_well_formatted_messages[string] PASSED [ 71%]
tests/unit/lms/validation/_exceptions_test.py::TestLTIToolRedirect::test_it_requires_well_formatted_messages[list] PASSED [ 85%]
tests/unit/lms/validation/_exceptions_test.py::TestLTIToolRedirect::test_it_requires_well_formatted_messages[dict of string] PASSED [100%]
...
```